cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/DistributedStreamService.java

```diff
@@ -107,6 +110,7 @@
   private LeaderElection leaderElection;
   private ResourceCoordinator resourceCoordinator;
   private Cancellable coordinationSubscription;
+  private ExecutorService heartbeatsSubscriptionExecutor;
```

Comment:
With @anwar6953 we saw in a test that he was writing for the stream notification logic, that when a heartbeat is published, it can trigger the publishing of a stream notification. If we don't specify another executor, then the same thread executor will be used, and when the second publishing occurs, it tries to submit a new publishing job, but the same thread executor is already busy publishing the heartbeat... So the executor will never be released and we enter an endless wait.